### DISCUSSION TITLE
Block Hankel Matrix Class

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1537] [Show and tell]
This is a rudimentary implementation of a `BlockHankelNumpyMatrixOperator` class I have written. It only keeps the necessary Hankel blocks in memory and does not construct the full matrix. Matrix-vector multiplications are realized by FFT-accelerated `apply` and `apply_adjoint` methods that work with the Hankel blocks directly.

I need this for efficient implementation of the Eigensystem Realization Algorithm where constructing the Hankel matrix for large data can become problematic quite quickly and efficient matrix-vector multiplications are needed to accelerate usage with a random range finder in RSVD. I suspect that having such a functionality in pymor could be advantageous for other methods that involve the Hankel matrix in the future.

There are plenty of questions that I hope can be discussed here in order to make this into a manageable PR some time. (Please add more questions):
- what is a good name for this? `BlockHankelNumpyMatrixOperator` isn't it. Moreover, this needs better names of variables in general.
- the circulant vector only needs to be computed once and should therefore be cached. How can this be implemented nicely? Should the class inherit from both `NumpyMatrixOperator` and `CacheableObject`?
- this implementation aims at scenarios where constructing the matrix explicitly is infeasible. How should this be handled in regards to the `assemble` methods (similarly methods that involve the inverse)?

### Circulant Matrix-Vector Multiplications

Let <img src="https://render.githubusercontent.com/render/math?math=c\in\mathbb{R}^n">, let <img src="https://render.githubusercontent.com/render/math?math=C"> denote the circulant matrix with first column <img src="https://render.githubusercontent.com/render/math?math=c"> and let <img src="https://render.githubusercontent.com/render/math?math=F_n"> denote the Fourier matrix of size <img src="https://render.githubusercontent.com/render/math?math=n">.

The eigenvalue decomposition of a circulant matrix <img src="https://render.githubusercontent.com/render/math?math=C"> is given by

<img src="https://render.githubusercontent.com/render/math?math=C=F_n^H\text{diag}(\lambda_1(C),\,\dots,\,\lambda_n(C))F_n">, where <img src="https://render.githubusercontent.com/render/math?math=[\lambda_1(C),\,\dots,\,\lambda_n(C)]^T=F_nc">.

Therefore, a matrix-vector multiplication can be efficiently realized with FFT by
<img src="https://render.githubusercontent.com/render/math?math=Cx=F_n^H(F_nc\odot F_nx)=IDFT(DFT(c)\odot DFT(x))">, where <img src="https://render.githubusercontent.com/render/math?math=\odot"> denotes the elementwise vector multiplication.

### Hankel Matrix-Vector Multiplications

Let <img src="https://render.githubusercontent.com/render/math?math=H\in\mathbb{R}^{n\times n}"> denote a Hankel matrix with entries <img src="https://render.githubusercontent.com/render/math?math=h\in\mathbb{R}^{2n}"> and let <img src="https://render.githubusercontent.com/render/math?math=J"> denote the exchange matrix of size <img src="https://render.githubusercontent.com/render/math?math=n"> (ones on the anti-diagonal). 

Since <img src="https://render.githubusercontent.com/render/math?math=Hx=HJJx">, we can rewrite it as
<img src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Cbegin%7Balign%2A%7DC%5Cbegin%7Bbmatrix%7DJx%5C%5C0%5Cend%7Bbmatrix%7D%3D%5Cbegin%7Bbmatrix%7DHJ%26X%5C%5CX%26HJ%5Cend%7Bbmatrix%7D%0A%5Cbegin%7Bbmatrix%7DJx%5C%5C0%5Cend%7Bbmatrix%7D%3D%0A%5Cbegin%7Bbmatrix%7Dy%5C%5C%5Cast%5Cend%7Bbmatrix%7D%0A%5Cend%7Balign%2A%7D">, where <img src="https://render.githubusercontent.com/render/math?math=\ast"> is the part of the solution that can be neglected.

The first column of <img src="https://render.githubusercontent.com/render/math?math=C"> is given by
<img src="https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+c%3D%5Cbegin%7Bbmatrix%7Dh_n+%26+h_%7Bn%2B1%7D+%26+%5Ccdots+%26+h_%7B2n-1%7D+%26+0+%26+h_1+%26+%5Ccdots+%26+h_%7Bn-1%7D%5Cend%7Bbmatrix%7D%5ET%5Cin%5Cmathbb%7BR%7D%5E%7B2n%7D">.

This means that <img src="https://render.githubusercontent.com/render/math?math=Hx"> can be efficiently realized by zero-padding <img src="https://render.githubusercontent.com/render/math?math=Jx"> with <img src="https://render.githubusercontent.com/render/math?math=n"> zeroes, applying circulant matrix vector multiplication with <img src="https://render.githubusercontent.com/render/math?math=c"> as above and discarding the last <img src="https://render.githubusercontent.com/render/math?math=n"> entries of the solution.

### Block Hankel Matrix-Vector Multiplications

The above strategy can easily be expanded to block Hankel matrices by performing the multiplication for every block entry individually (with some index picking drudgery).

### Python Testbench
Below, the `_circulant` is computed in the `__init__`. Ideally, I would like to have it as a cached property. `p`,`m` are the number of rows and columns of the blocks, respectively. `n` or `n_blocks` is the number of block matrices from which the Hankel matrix is constructed, whereas `s` is the "size of the Hankel matrix", i.e. the number of `(p,m)` blocks in every row/column.

```python
import numpy as np
import numpy.fft as nfft
from math import floor
from numpy.testing import assert_almost_equal
import matplotlib.pyplot as plt

from pymor.models.iosys import LTIModel
from pymor.operators.numpy import NumpyMatrixBasedOperator, NumpyMatrixOperator
from pymor.vectorarrays.numpy import NumpyVectorSpace


class BlockHankelNumpyMatrixOperator(NumpyMatrixBasedOperator):
    def __init__(
        self, matrix, source_id=None, range_id=None, solver_options=None, name=None
    ):
        assert matrix.ndim <= 3
        try:
            matrix.setflags(write=False)  # make numpy arrays read-only
        except AttributeError:
            pass

        p, m, s = matrix.shape
        if s % 2:  # add zero-block if n_block is uneven
            matrix = np.concatenate([matrix, np.zeros([p, m, 1])], axis=-1)
            s += 1

        self.__auto_init(locals())
        self.n_blocks = s // 2
        self.source = NumpyVectorSpace(m * self.n_blocks, source_id)
        self.range = NumpyVectorSpace(p * self.n_blocks, range_id)
        self._circulant = self._calc_circulant()

    def _assemble(self):
        pass

    def apply(self, U, mu=None):
        assert U in self.source
        U = U.to_numpy().T
        k = U.shape[1]
        n = self.n_blocks
        p, m, _ = self.matrix.shape
        y = self.range.zeros(k).to_numpy().T
        c = self._circulant
        for i in range(p):
            for j in range(m):
                x = np.concatenate([np.flip(U[j::m], axis=0), np.zeros([n, k])])
                y[i::p] += np.real(
                    nfft.ifft(nfft.fft(x, axis=0,) * c[i, j].reshape(-1, 1), axis=0,)
                )[:n]

        return self.range.make_array(y.T)

    def apply_adjoint(self, V, mu=None):
        assert V in self.range
        return self.H.apply(V, mu=mu)

    def _calc_circulant(self):
        p, m, _ = self.matrix.shape
        return nfft.fft(
            np.roll(
                np.concatenate([np.zeros([p, m, 1]), self.matrix[..., :-1]], axis=-1),
                self.n_blocks,
                axis=-1,
            )
        )

    @property
    def H(self):
        if np.isrealobj(self.matrix):
            adjoint_matrix = self.matrix.transpose(1, 0, 2)
        else:
            adjoint_matrix = self.matrix.transpose(1, 0, 2).conj()
        return self.with_(
            matrix=adjoint_matrix,
            source_id=self.range_id,
            range_id=self.source_id,
            name=self.name + "_adjoint",
        )

def make_hankel_blocks(s, p=2, m=3):
    return np.arange(p * m * s).reshape(p, m, s)

def construct_hankel_matrix_explicitly(h):
    p, m, s = h.shape
    if s % 2:  # add zero block, if number of blocks is uneven
        s += 1

    n = s // 2
    a = np.zeros([p * n, m * n])
    hz = np.concatenate([h, np.zeros([p, m, n])], axis=-1)
    # construct full block hankel matrix by hand
    for i in range(p * n):
        for j in range(m * n):
            a[i, j] = hz[i % p, j % m, floor(i / p) + floor(j / m)]

    return NumpyMatrixOperator(a)

# test apply methods
p, m, s = 2, 3, 12
h = make_hankel_blocks(s, p=p, m=m)
H = BlockHankelNumpyMatrixOperator(h)
A = construct_hankel_matrix_explicitly(h)
assert A.source == H.source
assert A.range == H.range
x = H.source.random(10)
assert_almost_equal(H.apply(x).to_numpy(), A.apply(x).to_numpy())
y = H.range.random(10)
assert_almost_equal(H.apply_adjoint(y).to_numpy(), A.apply_adjoint(y).to_numpy())
```
# Benchmarks
These runtime benchmarks should be taken with a grain of salt. I did not have time to dive into the details of my `numpy` config and I suspect multi-threading is active for `numpy.dot`. 
```python
# runtime plot
import perfplot

def make_A(n):
    return construct_hankel_matrix_explicitly(make_hankel_blocks(n))

def make_H(n):
    return BlockHankelNumpyMatrixOperator(make_hankel_blocks(n))

max_n_blocks = 10000
n_range = np.round(np.geomspace(5, max_n_blocks, 10)).astype(int)
out1 = perfplot.bench(
    setup=lambda n: (make_A(n), make_H(n), make_H(n).source.random()),
    kernels=[lambda x: x[0].apply(x[2]), lambda x: x[1].apply(x[2])],
    labels=["Regular", "Block-Hankel"],
    n_range=n_range,
    xlabel="Number of Hankel blocks",
    equality_check=None,
)

plot_opts = dict(logx=True)
out1.show(**plot_opts)
```
![runtime](https://user-images.githubusercontent.com/23401237/149520901-82783c64-1412-4dd0-8cad-a4efb458a990.png)

```python
# memory usage plot
size_A = np.zeros_like(n_range)
size_H = np.zeros_like(n_range)
for i, n in enumerate(n_range):
    size_A[i] = make_A(n).matrix.nbytes * 1e-6
    size_H[i] = make_H(n).matrix.nbytes * 1e-6

out2, ax = plt.subplots()
ax.semilogx(n_range, size_A, label="Regular")
ax.semilogx(n_range, size_H, label="Block-Hankel")
ax.set_xlabel("Number of Hankel blocks")
ax.set_ylabel("Size [Mb]")
plt.legend()
out2.show()
```
![memory](https://user-images.githubusercontent.com/23401237/149520925-3b8e4d7b-a5ec-4e36-83dd-66c1181486d0.png)

**Comment by sdrave** [comment]
Hi @artpelling, if I understand this correctly, a possible implementation might be via a `BlockOperator` of `HankelNumpyMatrixOperator`? Have you considered this? Since you have the double Python loop in `apply` anyway, it might be interesting to see how large the overhead of the pyMOR abstractions in your use case is.

> what is a good name for this? BlockHankelNumpyMatrixOperator isn't it. Moreover, this needs better names of variables in general.

It's a bit long, but I could very well live with that.

> the circulant vector only needs to be computed once and should therefore be cached. How can this be implemented nicely? Should the class inherit from both NumpyMatrixOperator and CacheableObject?

For this case I think it's fine to just use a private attribute as you are already doing. If you don't want to do the work in `__init__`, you can also later check in `apply` if `self._circulant` is defined. (While Operators a immutable, it is still allowed to write to private attributes after `__init__`. Of course, if you should be careful what you are doing ..)

> this implementation aims at scenarios where constructing the matrix explicitly is infeasible. How should this be handled in regards to the assemble methods (similarly methods that involve the inverse)?

Just don't inherit from `NumpyMatrixBasedOperator` and don't implement `assemble`. You also don't have to implement `apply_inverse`.

**Reply by artpelling** [reply]
> Hi @artpelling, if I understand this correctly, a possible implementation might be via a `BlockOperator` of `HankelNumpyMatrixOperator`? Have you considered this? Since you have the double Python loop in `apply` anyway, it might be interesting to see how large the overhead of the pyMOR abstractions in your use case is.

Indeed, this looks like a great way. I am about to write up a new example. However, I am puzzled by the code in `BlockOperator`. In the first line of the `__init__`, `self.blocks` is set to `np.array(blocks)`, meaning that the attribute will always be an `ndarray`. Later in the `__init__` it is asserted that the block entries are instances of `Operator`, but `self._operator` just yields the entries of the blocks...
 
I feel like I am missing pymor's type-paradigm in some way, as I am struggling with some other cases as well. I was not sure where to ask this so I would like to take this opportunity and ask about some other (less related) typing issues I have encountered. Maybe you (or somebody else) could briefly explain the reasoning behind the choices of types. 

- `svd_va`: The SVD methods take `VectorArray` as inputs and return a `VectorArray` U and `ndarray` Vh. Why are the return types of U and Vh different? Why can't I plug an `Operator` (or at least `NumpyMatrixOperator`) into the SVD routines?
- `lyapunov`: `solve_lyap_dense` takes `NumpyArray` (although the docstring also says `Operator` for E) and return `NumpyArray`, whereas `solve_lyap_lrcf` takes `Operator` inputs and returns `VectorArray`.

I have been learning pymor the last 6 weeks or so. While there are plenty of good tutorials and thorough documentation available, some basic ideas are not so obvious (at least to me). I feel like improving the docs in that area could lower the bar for new contributors substantially. I'm thinking something like an FAQ-style section about "pymor-ic" coding style. E.g. I still have not found the correct "pymor-ic" way to get the transpose of a `VectorArray` :sweat_smile:

**Reply by sdrave** [reply]
> > Hi @artpelling, if I understand this correctly, a possible implementation might be via a `BlockOperator` of `HankelNumpyMatrixOperator`? Have you considered this? Since you have the double Python loop in `apply` anyway, it might be interesting to see how large the overhead of the pyMOR abstractions in your use case is.
> 
> Indeed, this looks like a great way. I am about to write up a new example. However, I am puzzled by the code in `BlockOperator`. In the first line of the `__init__`, `self.blocks` is set to `np.array(blocks)`, meaning that the attribute will always be an `ndarray`. Later in the `__init__` it is asserted that the block entries are instances of `Operator`, but `self._operator` just yields the entries of the blocks...

`self._operators` is just an internal helper to iterator over all `Operators` in `self.blocks`. I agree that something like `self._iter_blocks` might make more sense ...


> I feel like I am missing pymor's type-paradigm in some way, as I am struggling with some other cases as well. I was not sure where to ask this so I would like to take this opportunity and ask about some other (less related) typing issues I have encountered. Maybe you (or somebody else) could briefly explain the reasoning behind the choices of types.

It is very important to understand and maybe not obvious for people with a numerical linear algebra background that `VectorArrays` are really lists of vector objects and not some glorified (transposed) matrix. When used with an external PDE solver, one typically uses `ListVectorArrays` which actually are implemented as Python lists of vector objects, and the vector objects correspond to internal data structures of the solver. It is generally assumed that the dimension the vectors can be very large, the object might be MPI distributed or actually only implicitly given as some function over a domain. This means that you cannot really access the 'data inside the vector' because it might be too large to handle or actually not even be there. There are exemptions to this rule, but the vast majority of algorithms in pyMOR assume that the only things you can do with vectors is add them, scale them or take inner products. 

On the other hand, it is assumed that the number of vectors in an array is relatively small (<1000, say), so we can handle scalar data associated with each vector in an array. E.g., we can compute the matrix of all inner products between vectors in the array or their norms. Since this data is small and will never be managed by the PDE solver, these quantities are returned from the interface methods as numpy arrays.

>     * `svd_va`: The SVD methods take `VectorArray` as inputs and return a `VectorArray` U and `ndarray` Vh. Why are the return types of U and Vh different? 

To answer this, we first need to be aware of the linear map of which the SVD is computed. In the case of a `VectorArray` of length N it is the linear map sending the first canonical basis vector of R^N to the first vector in the array and so on. So the source of the map is a small euclidean space whereas the range is some large vector space implemented in the solver, the elements of which we can only manipulate by means of `VectorArray` operations. So the left-singular vectors belonging to this space must be stored in a `VectorArray` U. On the other hand, the right-singular vectors in R^N can never live in an external solver, so it's fine to store them as a numpy array. To make this qualitative difference clear in the API we in particular chose numpy arrays over numpy arrays wrapped as `NumpyVectorArrays`.

>    Why can't I plug an `Operator` (or at least `NumpyMatrixOperator`) into the SVD routines?

Because the algorithms in `svd_va` operate on `VectorArrays`. Each `VectorArray` can be interpreted as a linear `Operator` (also see `VectorArrayOperator`), but you generally cannot convert an `Operator` to a `VectorArrray` because both the range and the source of the `Operator` might live in an external solver. Also, the operator might be implemented matrix-free.

Unfortunately, there are currently no SVD algorithms for `Operators` in pyMOR. A randomized SVD would surely accept `Operators` and maybe @lbalicki will sometime implement something in line of `pymor.algorithms.eigs` ..


>     * `lyapunov`: `solve_lyap_dense` takes `NumpyArray` (although the docstring also says `Operator` for E) and return `NumpyArray`, whereas `solve_lyap_lrcf` takes `Operator` inputs and returns `VectorArray`.

This is because the dense methods, unfortunately, cannot work with abstract pyMOR `Operators`. Thus, using pyMOR abstractions here, would offer no real advantage and might mislead the user to believe these methods could work with arbitrary `Operators`. (Regarding `E`, the docstring clearly seems wrong. Could you open an issue?)

> I have been learning pymor the last 6 weeks or so. While there are plenty of good tutorials and thorough documentation available, some basic ideas are not so obvious (at least to me). I feel like improving the docs in that area could lower the bar for new contributors substantially. 

I agree that many newcomers have some issues with the `VectorArray` concept. I have been planning on writing a `VectorArray` tutorial for quite some time, but never got to it, unfortunately. Hopefully for the next release ..

> I'm thinking something like an FAQ-style section about "pymor-ic" coding style. 

I believe we decided against that a while ago, but maybe we should think about it again. However, there is also the Q&A section in Github Discussions which can serve a similar purpose. I have put it in the agenda of our next dev meeting ..

> E.g. I still have not found the correct "pymor-ic" way to get the transpose of a `VectorArray` sweat_smile

The transpose of a `VectorArray` does not really make sense. But in almost all situations what you want to do will either be achievable with `inner` or `lincomb` (see `VectorArrayOperator` with `adjoint=True/False`).

**Reply by artpelling** [reply]
Thank you very much for the long reply, this really clarified a lot of things! 

> It is very important to understand and maybe not obvious for people with a numerical linear algebra background that VectorArrays are really lists of vector objects and not some glorified (transposed) matrix.

I was reading the Technical Overview section again with this in mind and everything you said is sort of contained in it. The implications were just not immediately clear to me (I am lacking intuition when it comes to external PDE solvers and indeed come from a numerical linear algebra / numpy backgroud).

As I see that there are plans to integrate data-driven MOR methods, I suspect that similar topics will come up again. I guess the exciting work for the future will be how to marry those two areas. Or conversely, where to draw the line and keep them separate... 

> This is because the dense methods, unfortunately, cannot work with abstract pyMOR Operators. Thus, using pyMOR abstractions here, would offer no real advantage and might mislead the user to believe these methods could work with arbitrary Operators. (Regarding E, the docstring clearly seems wrong. Could you open an issue?)

Done #1541 #1542 

> self._operators is just an internal helper to iterator over all Operators in self.blocks. I agree that something like self._iter_blocks might make more sense ...

I understand, but I am still puzzled. From what I see, `self.blocks` are cast to numpy arrays in

https://github.com/pymor/pymor/blob/eaf4dfb29e340d373d13a470d07d91e8fa78b2e9/src/pymor/operators/block.py#L12-L20

How can the blocks then be Operators? I read up on double assignments - I understand that `self.blocks` and `blocks` are both casted numpy arrays..

**Reply by sdrave** [reply]
> I was reading the Technical Overview section again with this in mind and everything you said is sort of contained in it. The implications were just not immediately clear to me (I am lacking intuition when it comes to external PDE solvers and indeed come from a numerical linear algebra / numpy backgroud).

Yes, I think many people are in a similar situation.

> As I see that there are plans to integrate data-driven MOR methods, I suspect that similar topics will come up again. I guess the exciting work for the future will be how to marry those two areas. Or conversely, where to draw the line and keep them separate...

Yeah, we already had several discussion in that direction. #1350 can be seen as a first step towards separating algorithms which need access to vector data from algorithms which don't.

> > This is because the dense methods, unfortunately, cannot work with abstract pyMOR Operators. Thus, using pyMOR abstractions here, would offer no real advantage and might mislead the user to believe these methods could work with arbitrary Operators. (Regarding E, the docstring clearly seems wrong. Could you open an issue?)
> 
> Done #1541 #1542

Thanks!


> 
> > self._operators is just an internal helper to iterator over all Operators in self.blocks. I agree that something like self._iter_blocks might make more sense ...
> 
> I understand, but I am still puzzled. From what I see, `self.blocks` are cast to numpy arrays in
> 
> https://github.com/pymor/pymor/blob/eaf4dfb29e340d373d13a470d07d91e8fa78b2e9/src/pymor/operators/block.py#L12-L20
> 
> How can the blocks then be Operators? I read up on double assignments - I understand that `self.blocks` and `blocks` are both casted numpy arrays..

Now I understand the confusion: `self.block` is a numpy array of `dtype` `object`. Every array element is still an `Operator`.

**Reply by artpelling** [reply]
Thanks for taking the time! #1350 looks very interesting. I will go ahead and write up that blocked version that you suggested :)

**Reply by artpelling** [reply]
Here are some perfomance comparisons. I have chosen different block sizes.
`p=m=2`:
![plot_2x2](https://user-images.githubusercontent.com/23401237/150821850-7a9f5bab-ba7b-4a7c-b5b5-158371b9ce7f.png)
and `p=m=100`:
![plot_100x100](https://user-images.githubusercontent.com/23401237/150821854-bec16d0e-17c2-417e-a891-8d7cdc8640c8.png)

Since this class is aimed at large operators, the additional overhead seems acceptable to me.
The code of the new class I used for these test can be found in #1546. I suggest to move the discussion over to the PR.

**Reply by sdrave** [reply]
I would agree that the overhead is acceptable. However, what I don't get is why the pyMOR implementation looks worse for a small number of blocks and large block size compared to a small number of blocks and small block size. I would have expected the opposite. Do you understand why this happens?
Also, you might want to check if running with `python -O` (which disables all assertions) has a relevant effect.

**Reply by artpelling** [reply]
> I would agree that the overhead is acceptable. However, what I don't get is why the pyMOR implementation looks worse for a small number of blocks and large block size compared to a small number of blocks and small block size. I would have expected the opposite. Do you understand why this happens? 

I think that is expected behaviour. Maybe the terminology here is a bit confusing? If we have a block Hankel matrix of dimension, say `ps,ms`, then the number of blocks `2s-1` should not have any influence on the runtime difference of both variants as it just scales up the FFT operations and spectral multiplications which are implemented identically. The only difference is how the individual block entries are looped over. Hence the block-size (i.e. `p` and `m`) should affect the runtime difference. As the sizes of the individual blocks stay constant for each plot, the runtime difference should also be somewhat constant. This leads to a decreasing gap in both curves, if the number of Hankel blocks, `2s-1`, is increased (as can be observed).

> Also, you might want to check if running with `python -O` (which disables all assertions) has a relevant effect.

Will do!

**Reply by artpelling** [reply]
The assertions seem to make up about half of the overhead for larger blocks.
`p=m=2` without assertions:
![plot_2x2_no_assert](https://user-images.githubusercontent.com/23401237/151043401-24b7718b-dbb0-4efe-aa53-f96f546481a2.png)
`p=m=100` without assertions:
![plot_100x100_no_assert](https://user-images.githubusercontent.com/23401237/151043435-b4d55c42-4661-4bd7-828e-32b946e1d0e8.png)
